pkg/river/token/builder/builder_test.go

```diff
@@ -409,3 +409,346 @@ func TestBuilder_MapBlocks(t *testing.T) {
+
+		BlockPtr:       &InnerBlockDefaulter{},
+		BlocksPtrSlice: []*InnerBlockDefaulter{},
+	})
+
+	expect := format(t, `
+		field = "some_value"
+
+		block_ptr { }
+	`)
+
+	require.Equal(t, expect, string(f.Bytes()))
+}
+
+func TestBuilder_AppendInner_Nil(t *testing.T) {
+	type Structure struct {
+		Field          string                 `river:"field,attr"`
+		Block          InnerBlockDefaulter    `river:"block,block,optional"`
+		BlocksSlice    []InnerBlockDefaulter  `river:"block_slice,block,optional"`
+		BlockPtr       *InnerBlockDefaulter   `river:"block_ptr,block,optional"`
+		BlocksPtrSlice []*InnerBlockDefaulter `river:"block_ptr_slice,block,optional"`
+	}
+
+	f := builder.NewFile()
+	f.Body().AppendFrom(Structure{
+		Field: "some_value",
+
+		Block:       InnerBlockDefaulter{},
+		BlocksSlice: nil,
+
+		BlockPtr:       nil,
+		BlocksPtrSlice: nil,
+	})
+
+	expect := format(t, `
+		field = "some_value"
+	`)
+
+	require.Equal(t, expect, string(f.Bytes()))
+}
+
+func TestBuilder_AppendInner_NilInSlices(t *testing.T) {
+	type Structure struct {
+		Field          string                 `river:"field,attr"`
+		Block          InnerBlockDefaulter    `river:"block,block,optional"`
+		BlocksSlice    []InnerBlockDefaulter  `river:"block_slice,block,optional"`
+		BlockPtr       *InnerBlockDefaulter   `river:"block_ptr,block,optional"`
+		BlocksPtrSlice []*InnerBlockDefaulter `river:"block_ptr_slice,block,optional"`
+	}
+
+	f := builder.NewFile()
+	f.Body().AppendFrom(Structure{
+		Field:          "some_value",
+		BlocksPtrSlice: []*InnerBlockDefaulter{nil, nil, nil},
+	})
+
+	expect := format(t, `
+		field = "some_value"
+	`)
+
+	require.Equal(t, expect, string(f.Bytes()))
+}
+
+func TestBuilder_AppendInnerBlock_InnerDefaultsMatch_OuterDefaultsNil(t *testing.T) {
+	f := builder.NewFile()
+
+	OuterBlockDefaults = OuterBlockDefaulter{
+		Inner:         InnerBlockDefaulter{},
+		InnerSlice:    nil,
+		InnerPtr:      nil,
+		InnerPtrSlice: nil,
+	}
+
+	outerUserValue := OuterBlockDefaulter{
+		Inner:         InnerDefaults,
+		InnerSlice:    []InnerBlockDefaulter{},
+		InnerPtr:      &InnerDefaults,
+		InnerPtrSlice: []*InnerBlockDefaulter{},
+	}
+
+	f.Body().AppendFrom(outerUserValue)
+
+	expect := format(t, `
+		inner { }
+		
+		inner_ptr { }
+	`)
+
+	require.Equal(t, expect, string(f.Bytes()))
+}
+
+func TestBuilder_AppendInnerAttr_InnerDefaultsMatch_OuterDefaultsNil(t *testing.T) {
+	f := builder.NewFile()
+
+	OuterAttrDefaults = OuterAttrDefaulter{
+		Inner:         InnerBlockDefaulter{},
+		InnerSlice:    nil,
+		InnerPtr:      nil,
+		InnerPtrSlice: nil,
+	}
+
+	outerUserValue := OuterAttrDefaulter{
+		Inner:         InnerDefaults,
+		InnerSlice:    []InnerBlockDefaulter{},
+		InnerPtr:      &InnerDefaults,
+		InnerPtrSlice: []*InnerBlockDefaulter{},
+	}
+
+	f.Body().AppendFrom(outerUserValue)
+
+	expect := format(t, `
+		inner = {
+			number = 123,
+		}
+		inner_slice = []
+		inner_ptr   = {
+			number = 123,
+		}
+		inner_ptr_slice = []
```

Comment:
This test's input data is virtually the same as the test above (`TestBuilder_AppendInnerBlock_InnerDefaultsMatch_OuterDefaultsNil`) with the only difference that instead of `block`, the `OuterAttrDefaulter` uses `attr` for its fields.

I would expect a result that is closer to how we handle a `block`, without printing default values for optional attributes?